kimchi/src/circuits/cairo/runner/memory.rs

```diff
@@ -0,0 +1,137 @@
+        // Safely convert idx from F to usize (since this is a memory address
+        // idx should not be too big, this should be safe)
+        let addr: u64 = idx.to_u64();
+        &self.data[addr as usize]
+    }
+}
+
+impl<F: PrimeField> IndexMut<F> for CairoMemory<F> {
+    fn index_mut(&mut self, idx: F) -> &mut Self::Output {
+        let addr: u64 = idx.to_u64();
+        &mut self.data[addr as usize]
+    }
+}
+
+impl<F: PrimeField> CairoMemory<F> {
+    /// Create a new memory structure from a vector of field elements
+    pub fn new(input: Vec<F>) -> CairoMemory<F> {
+        // Initialized with the public memory (compiled instructions only)
+        // starts intentionally with a zero word for ease of testing
+        let mut aux = vec![F::zero()];
+        aux.extend(input);
+        CairoMemory {
+            codelen: aux.len() - 1,
+            data: aux.into_iter().map(|i| Some(CairoWord::new(i))).collect(),
+        }
+    }
+
+    /// Get size of the public memory
+    pub fn get_codelen(&self) -> usize {
+        self.codelen
+    }
+
+    /// Get size of the full memory including dummy 0th entry
+    pub fn size(&self) -> u64 {
+        self.data.len() as u64
+    }
+
+    /// Resizes memory with enough additional None slots if necessary before writing or reading
+    fn resize(&mut self, addr: u64) {
+        // if you want to access an index of the memory but its size is less or equal than this
+        // you will need to extend the vector with enough spaces (taking into account that
+        // vectors start by index 0, the 0 address is dummy, and size starts in 1)
+        if let Some(additional) = addr.checked_sub(self.size() - 1) {
+            self.data.extend(repeat(None).take(additional as usize));
+        }
+    }
+
+    /// Write u64 element in memory address
+    pub fn write(&mut self, addr: F, elem: F) {
+        self.resize(addr.to_u64());
+        self[addr] = Some(CairoWord::new(elem));
+    }
+
+    /// Read element in memory address
+    /// Because of how assignments work in Cairo (called assert-equal), they
+    /// behave in two ways: either check two variables are equal or assign
+    /// the value of one variable to the address of the other one so that
+    /// their content will be the same. This means you may first read to
+    /// addresses of memory that were still not instantiated, and you need
+    /// to enlarge the vector before reading (with None values).
+    pub fn read(&mut self, addr: F) -> Option<F> {
+        self.resize(addr.to_u64());
+        if self[addr].is_some() {
+            Some(self[addr].unwrap().get_word())
+        } else {
+            None
+        }
+    }
+
+    /// Visualize content of memory excluding the 0th dummy entry
+    pub fn view(&mut self) {
+        for i in 1..self.size() {
+            if self.read(F::from(i)).is_some() {
+                println!("{}: 0x{:?}", i, self.read(F::from(i)).unwrap().to_hex());
```

Comment:
Oh sure, I'll do that! In any case, I think it is more intuitive if I don't use `to_hex()` here, but a little endian version of it. Otherwise things like `0x8000u16` get printed as `0x0080`.